cmd/kube-dns/SPEC.md

```diff
@@ -0,0 +1,99 @@
+
+Any DNS-based service discovery solution for Kubernetes must support the
+queries described below to be considered compliant with this specification.
+
+In the query names below, values not in angle brackets, `< >`,
+are literals. The meaning of the values in angle brackets are defined
+below or in the description of the specific query response.
+
+`<zone>` = configured cluster domain, e.g. cluster.local
+
+`<ns>` = a Namespace
+
+All comparisons between query names and data in Kubernetes is **case-insensitive**.
+
+In the results column, the following definitions should be used for
+words in _italics_.
+
+- _hostname_
+
+  - For endpoints, in order of precedence:
+    - The value of the endpoint’s `hostname` field.
+    - The value of the `HostName` field stored in the map indexed by the endpoint’s IP address in the `endpoints.beta.kubernetes.io/hostnames-map` annotation.
+    - A unique, system-assigned identifier for the endpoint. The exact format and source of this identifier is not prescribed by this specification. However, it must be possible to use this to identify a specific endpoint in the context of a Service. This is used for SRV queries below in the event no explicit endpoint hostname is defined by the above methods.
+
+- _ready_
+
+ - An endpoint is considered _ready_ if its address is in the `addresses` field of the EndpointSubset object, or the corresponding service has the `service.alpha.kubernetes.io/tolerate-unready-endpoints` annotation set to `true`.
+
+The supported queries and expected responses follow.
+
+- **Type:** A, **Name:** `<service>.<ns>.svc.<zone>`
+
+  - If no Service named `<service>` exists in `<ns>`, the result will be `NXDOMAIN`.
+
+  - If a Service named `<service>` exists in `<ns>` with a ClusterIP, then the result will be a single A record with the specified name and the ClusterIP of that Service.
+
+  - If a Service named `<service>` exists in `<ns>` without a ClusterIP (i.e., a headless service)  and there are no _ready_ endpoints for that Service, then the result will be empty (i.e., an empty Answer section).
+
+  - If a Service named `<service>` exists in `<ns>` without a ClusterIP and there are _ready_ endpoints for the Service, the result will be an A record with the specified name and IP for each _ready_ endpoint.
+
+- **Type:** A, **Name:** `<hostname>.<service>.<ns>.svc.<zone>`
+
+  - If no Service named `<service>` exists in `<ns>`, the result will be `NXDOMAIN`.
+
+  - If a Service named `<service>` exists in `<ns>` and a _ready_ endpoint exists for that Service with _hostname_ of `<hostname>` the result will be a single A record with the IP address of that endpoint. If no such endpoint exists, the result will be `NXDOMAIN`.
+
+- **Type:** A, **Name:** `<a>-<b>-<c>-<d>.<ns>.pod.<zone>`
+
+  - If a Pod exists with IP address `<a>.<b>.<c>.<d>`, the result will be an A record with the specified name and the IP address `<a>.<b>.<c>.<d>`.
+
+  - If no such Pod exists, the result will be `NXDOMAIN`.
```

Comment:
@thockin I believe that it is important to respond based on the state of the cluster. My understanding of the use case for the existing `<ip>.pod.cluster.local` responses is to enable a wildcard certificate like *.ns.pod.cluster.local. However, if the DNS always returns the IP regardless of the state of the cluster, then you are weakening the guarantee of identify provided by certificate.

In cases where a cert is being used like this, there must be an internal application-level discovery mechanism being used to tell clients about the various pods that are running as servers. This is the only way the clients would know about the DNS name like `a-b-c-d.default.pod.cluster.local`. So, consider the situation in which someone is able to subvert that discovery mechanism. Suppose I have a namespace, `goodguy` with a set of pods and I get a signed wildcard certificate `*.goodguy.pod.cluster.local`. Now, there is someone nefarious on my cluster who launches a malicious pod with IP w.x.y.z in a different namespace (say `badguy`). If the DNS reflects the real state of the cluster, then the bad guy would have to push my IP into the discovery mechanism as `w-x-y-z.badguy.pod.cluster.local`, and so the certificate validation would fail, and our clients would ignore that pod. However, if the DNS always echoes back the IP without validating the cluster state, then the bad guy can push `w-x-y-z.goodguy.pod.cluster.local` into the discovery mechanism, and still have the clients reach the malicious pod. The certificate will be seen as valid and matching the server host name.